enterprise/models.py

```diff
@@ -489,10 +579,25 @@ def enroll(self, course_run_id, mode):
                 'enterprise_customer_name': self.enterprise_customer.name,
                 'mode': mode,
             })
-        EnterpriseCourseEnrollment.objects.get_or_create(
-            enterprise_customer_user=self,
-            course_id=course_run_id
-        )
+            EnterpriseCourseEnrollment.objects.get_or_create(
+                enterprise_customer_user=self,
+                course_id=course_run_id
+            )
+        elif enrolled_in_course and course_enrollment.get('mode') in paid_modes and mode in audit_modes:
+            # This enrollment is attempting to "downgrade" the user from a paid track they are already in.
+            raise UserAlreadyEnrolledInCourse(
+                'The user is already enrolled in the course {course_run_id} in {current_mode} mode '
+                'and cannot be enrolled in {given_mode} mode'.format(
+                    course_run_id=course_run_id,
+                    current_mode=course_enrollment.get('mode'),
+                    given_mode=mode,
+                )
+            )
+        else:
+            # The user was otherwise already enrolled.
+            raise UserAlreadyEnrolledInCourse('The user is already enrolled in the course {course_run_id}'.format(
+                course_run_id=course_run_id
+            ))
```

Comment:
Do we need to return an error for these conditions where the learner is already enrolled? I'm not sure what the caller of the API endpoint would gain from receiving this information.

Reply:
I guess I'm erring on the side of being explicit, just to say that no action was performed because they were already enrolled, or the other case which was they were attempting to "downgrade" the enrollment. I could still catch this exception at the top level view and return success, but I like having the exception because it will short circuit the rest of the logic that doesn't need to be performed.

Reply:
Okay, just seems like a lot of extra code and logic paths for no real value.

Reply:
I see where you're coming from, and I don't disagree. I think the only case that is really an error according to business logic is the downgrade case - it seems like the caller should know why the enrollment is in a different mode than what they specified in their call, or that they shouldn't expect switching from verified to audit to work if they tried purposefully. Or maybe it's something we should allow?

Reply:
We should definitely prevent the downgrade and I agree that we should return an error status and message as the response to a request that attempts to do so.